core/templates/dev/head/components/state-editor/state-editor-properties-services/state-property.service.ts

```diff
@@ -16,74 +16,101 @@
-    // maintained if e.g. the state is renamed.
-    return {
-      init: function(stateName, value) {
-        if (this.setterMethodKey === null) {
-          throw 'State property setter method key cannot be null.';
-        }
-
-        // The name of the state.
-        this.stateName = stateName;
-        // The current value of the property (which may not have been saved to
-        // the frontend yet). In general, this will be bound directly to the UI.
-        this.displayed = angular.copy(value);
-        // The previous (saved-in-the-frontend) value of the property. Here,
-        // 'saved' means that this is the latest value of the property as
-        // determined by the frontend change list.
-        this.savedMemento = angular.copy(value);
-      },
-      // Returns whether the current value has changed from the memento.
-      hasChanged: function() {
-        return !angular.equals(this.savedMemento, this.displayed);
-      },
-      // The name of the setter method in ExplorationStatesService for this
-      // property. THIS MUST BE SPECIFIED BY SUBCLASSES.
-      setterMethodKey: null,
-      // Transforms the given value into a normalized form. THIS CAN BE
-      // OVERRIDDEN BY SUBCLASSES. The default behavior is to do nothing.
-      _normalize: function(value) {
-        return value;
-      },
-      // Validates the given value and returns a boolean stating whether it
-      // is valid or not. THIS CAN BE OVERRIDDEN BY SUBCLASSES. The default
-      // behavior is to always return true.
-      _isValid: function(value) {
-        return true;
-      },
-      // Updates the memento to the displayed value.
-      saveDisplayedValue: function() {
-        if (this.setterMethodKey === null) {
-          throw 'State property setter method key cannot be null.';
-        }
-
-        this.displayed = this._normalize(this.displayed);
-        if (!this._isValid(this.displayed) || !this.hasChanged()) {
-          this.restoreFromMemento();
-          return;
-        }
-
-        if (angular.equals(this.displayed, this.savedMemento)) {
-          return;
-        }
-
-        AlertsService.clearWarnings();
-
-        this.savedMemento = angular.copy(this.displayed);
-      },
-      // Reverts the displayed value to the saved memento.
-      restoreFromMemento: function() {
-        this.displayed = angular.copy(this.savedMemento);
-      }
-    };
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import cloneDeep from 'lodash/cloneDeep';
+
+import { AlertsService } from 'services/alerts.service';
+import { UtilsService } from 'services/utils.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class StatePropertyService {
+  // The name of the setter method in ExplorationStatesService for this
+  // property. THIS MUST BE SPECIFIED BY SUBCLASSES.
+  setterMethodKey: string;
+  displayed: any;
+  stateName: string;
+  savedMemento: any;
```

Comment:
What's the type for `displayed` and `savedMemento`?

Reply:
`displayed` and `savedMemento` both have the type as `value` of the property for which the service works. We need to have domain objects and assign type accordingly then. I have added a TODO now.